### PR TITLE
[DWARFDump] Make --verify handle all sections by default

### PR DESCRIPTION
**Pull request by felipepiovezan**
The current behavior of --verify is that it only verifies debug_info, debug_abbrev and debug_names. This seems fairly arbitrary and might have been unintentional, as originally the absence of any section flags implied "all".

This patch changes the behavior so that the verifier now verifies everything by default. It revealed two tests that had potentially invalid DWARF:

1. dwarfdump-str-offsets.s is adding padding between two debug_str_offset contributions. The standard does not explicitly allow this behavior. See issue https://github.com/llvm/llvm-project/issues/81558

2. dwarf5-macro.test uses a checked-in binary that has invalid debug_str_offsets. One of its entries points to the _middle_ of the string section:

error: .debug_str_offsets: contribution 0x0: index 0x4: invalid string offset *0x18 == 0x455D, is neither zero nor immediately following a null character

If we look at the closest offset to 0x455D in debug_str:

```
0x0000454e: "__SLONG32_TYPE int"
```

0x455D points to "int".

### CODE REVIEW

**Comment by dwblaikie** [comment]
> ```
> 0x0000454e: "__SLONG32_TYPE int"
> ```
> 
> 
>     
>   
> 
> 0x455D points to "int".

It's possible that this is valid DWARF and the verifier is being overly pedantic. This would be a tail merging operation - I don't think LLVM does this, but it's possible that other producers do?

File: llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp [hunk 1]
```diff
@@ -866,7 +866,8 @@ int main(int argc, char **argv) {
     if (Verbose)
       DumpType = DIDT_All;
     else
-      DumpType = DIDT_DebugInfo;
+      // If no options are passed, verify everything but dump only debug_info.
```
Comment (JDevlieghere) [inline llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:869]: Nit: this comment is slightly ambiguous. I think you can sidestep it by using an `else if` so it looks symmetrical with the verbose dumping. 

```
if (Verbose)
  DumpType = DIDT_All;
else if(Verify)
  DumpType = DIDT_All;
else 
  DumpType = DIDT_DebugInfo;
```

or maybe even simpler:

```
if (Verbose || Verify)
  DumpType = DIDT_All;
else 
  DumpType = DIDT_All;
```

**Review by JDevlieghere — APPROVED** [review]
LGTM

**Comment by felipepiovezan** [comment]
> It's possible that this is valid DWARF and the verifier is being overly pedantic. This would be a tail merging operation - I don't think LLVM does this, but it's possible that other producers do?

The spec doesn't say anything, only that strps must point to some offset inside the string table. If we interpret this literally, it would be valid DWARF indeed. I see no reason to disallow that 🤔 

Looking into the commit history and... you're the one who implemented this check! :) You actually point this out in the commit description (5e74b2e8bb9b191aa5ed433820860b30ca7d9baa):

>  This assumes no suffix merging - if anyone implements that, then this
    checking should just be removed for the most part (we could still check
    the offsets are within the bounds of .debug_str[.dwo], but nothing more
    - any offset in the range would be valid, the offsets wouldn't have to
    land at the start of a string)

It curious because the affected test -- if we are to believe the comments there -- was generated by Clang... I tried to generate its binary inputs again, but found a crash in Clang's handling of debug macros, so I decided to put it off for now and just disable the verifier for that check; this should be fine, the test is not focused in debug strings.

File: llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp (hunk 1 above)
Comment (felipepiovezan) [reply llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:869]: Yup, good point! I like your second suggestion.

**Review by avl-llvm — APPROVED** [review]

**Review by clayborg — APPROVED** [review]

**Pull request merged**